File: Codinator/Other Sources/Syntax/JS/JsTextView.m
//
//  JsTextView.m
//  VWAS-HTML
//
//  Created by Vladimir on 20/03/15.
//  Copyright (c) 2015 Vladimir Danila. All rights reserved.
//


@import CoreText;
#import "JsTextView.h"


#define RGB(r,g,b) [UIColor colorWithRed:r/255.0f green:g/255.0f blue:b/255.0f alpha:1.0f]

@implementation JsTextView

#pragma mark - Initialization & Setup

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    
    if (self)
    {
        [self commonSetup];
    }
    
    return self;
}

- (void)commonSetup
{
    size = [[[NSUserDefaults standardUserDefaults] objectForKey:@"fontSize"] integerValue];
    if (size == 0) {
        
        size = 14;
        
        [[NSUserDefaults standardUserDefaults] setObject:[NSNumber numberWithInteger:size] forKey:@"fontSize"];
        
    }
    
    
    _boldFont = [UIFont boldSystemFontOfSize:size];
    _italicFont = [UIFont italicSystemFontOfSize:size];
     self.font = _defaultFont;
    _defaultFont = [UIFont systemFontOfSize:size];

    
    
    NSString *DarkEq = [[NSUserDefaults standardUserDefaults] stringForKey:@"DarkMode"];
    NSLog(@"%@",DarkEq);
    if ([DarkEq isEqualToString:@"Ja"]) {
        self.textColor = [UIColor whiteColor];
        
    }
    else{
        self.textColor = [UIColor blackColor];
        
    }
    
    
    [self addObserver:self forKeyPath:NSStringFromSelector(@selector(defaultFont)) options:NSKeyValueObservingOptionNew context:0];
    [self addObserver:self forKeyPath:NSStringFromSelector(@selector(boldFont)) options:NSKeyValueObservingOptionNew context:0];
    [self addObserver:self forKeyPath:NSStringFromSelector(@selector(italicFont)) options:NSKeyValueObservingOptionNew context:0];
    
    if (_italicFont == nil && ([UIFontDescriptor class] != nil))
    {
        // This works around a bug in 7.0.3 where HelveticaNeue-Italic is not present as a UIFont option
        _italicFont = (__bridge_transfer UIFont*)CTFontCreateWithName(CFSTR("HelveticaNeue-Italic"), 14.0f, NULL);
    }
    
    self.tokens = [self solverTokens];
}


- (NSArray *)solverTokens
{
    NSArray *solverTokens =  @[
                               
                               [CYRToken tokenWithName:@"special_numbers"
                                            expression:@"[ʝ]"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : [UIColor orangeColor]
                                                         }],
                               [CYRToken tokenWithName:@"mod"
                                            expression:@"\bmod\b"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(245, 0, 110)
                                                         }],

                                [CYRToken tokenWithName:@"string"
                                            expression:@"\".*?(\"|$)"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(24, 110, 109)
                                                         }],
                               [CYRToken tokenWithName:@"string2"
                                            expression:@"'.*?('|$)"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(186, 114, 181),
                                                         }],
                               [CYRToken tokenWithName:@"hex_1"
                                            expression:@"\\$[\\d a-f]+"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(186, 114, 181),
                                                         }],
                               [CYRToken tokenWithName:@"octal_1"
                                            expression:@"&[0-7]+"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(186, 114, 181),
                                                         }],
                               [CYRToken tokenWithName:@"binary_1"
                                            expression:@"%[01]+"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(186, 114, 181),
                                                         }],
                               [CYRToken tokenWithName:@"hex_2"
                                            expression:@"0x[0-9 a-f]+"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(186, 114, 181),
                                                         }],
                               [CYRToken tokenWithName:@"octal_2"
                                            expression:@"0o[0-7]+"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(186, 114, 181),
                                                         }],
                               [CYRToken tokenWithName:@"binary_2"
                                            expression:@"0b[01]+"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(186, 114, 181),
                                                         }],
                               [CYRToken tokenWithName:@"float"
                                            expression:@"\\d+\\.?\\d+e[\\+\\-]?\\d+|\\d+\\.\\d+|∞"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(186, 114, 181),
                                                         }],
                               [CYRToken tokenWithName:@"integer"
                                            expression:@"\\d+"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(186, 114, 181), // purple
                                                         }],
                               [CYRToken tokenWithName:@"operator"
                                            expression:@"[/\\*,\\;:=<>\\+\\-\\^!·≤≥]"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(255, 255, 255) //white
                                                         }],
                               [CYRToken tokenWithName:@"round_brackets"
                                            expression:@"[\\(\\)]"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(255, 255, 255), //white

                                                         }],
                               [CYRToken tokenWithName:@"square_brackets"
                                            expression:@"[\\[\\]]"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(255, 255, 255), //white
                                                         }],
                               [CYRToken tokenWithName:@"absolute_brackets"
                                            expression:@"[|]"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(255, 255, 171)
                                                         }],
                               [CYRToken tokenWithName:@"reserved_words"
                                            expression:@"(abs|acos|acosh|function|asin|asinh|atan|atanh|atomicweight|ceil|complex|cos|cosh|crandom|deriv|erf|erfc|exp|eye|floor|frac|gamma|gaussel|getconst|imag|inf|integ|integhq|inv|ln|log10|log2|machineprecision|max|maximize|min|minimize|molecularweight|ncum|ones|pi|plot|random|real|round|sgn|sin|sqr|sinh|sqrt|tan|tanh|transpose|trunc|true|false|func|var|zeros|@import|void|return)"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(186, 114, 181), //purple
                                                         }],
                               [CYRToken tokenWithName:@"chart_parameters"
                                            expression:@"(chartheight|charttitle|chartwidth|color|seriesname|showlegend|showxmajorgrid|showxminorgrid|showymajorgrid|showyminorgrid|transparency|thickness|xautoscale|xaxisrange|xlabel|xlogscale|xrange|yautoscale|yaxisrange|ylabel|ylogscale|yrange)"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(186, 114, 181),
                                                         }],
                               [CYRToken tokenWithName:@"comment"
                                            expression:@"//.*"
                                            attributes:@{
                                                         NSForegroundColorAttributeName : RGB(31, 131, 0),
                                                         NSFontAttributeName : self.italicFont
                                                        }],



    
    
                               ];
    
    return solverTokens;
}


#pragma mark - Cleanup

- (void)dealloc
{
    [self removeObserver:self forKeyPath:NSStringFromSelector(@selector(defaultFont))];
    [self removeObserver:self forKeyPath:NSStringFromSelector(@selector(boldFont))];
    [self removeObserver:self forKeyPath:NSStringFromSelector(@selector(italicFont))];
}

#pragma mark - KVO

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
    if ([keyPath isEqualToString:NSStringFromSelector(@selector(defaultFont))] ||
        [keyPath isEqualToString:NSStringFromSelector(@selector(boldFont))] ||
        [keyPath isEqualToString:NSStringFromSelector(@selector(italicFont))])
    {
        // Reset the tokens, this will clear any existing formatting
        self.tokens = [self solverTokens];
    }
    else
    {
        //[super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
    }
}



@end

File: Codinator/Other Sources/Syntax/JS/JsTextView.swift
//
//  NeuronTextView.swift
//  Codinator
//
//  Created by Vladimir Danila on 28/03/16.
//  Copyright © 2016 Vladimir Danila. All rights reserved.
//

import UIKit

class JsTextView: CYRTextView {

    var defaultFont: UIFont = NSUserDefaults.standardUserDefaults().fontForKey("Font: 0") {
        didSet {
            tokens = highlightingTokens()
        }
    }
    
    var boldFont: UIFont = NSUserDefaults.standardUserDefaults().fontForKey("Font: 1") {
        didSet {
            tokens = highlightingTokens()
        }
    }
    
    var italicFont: UIFont = NSUserDefaults.standardUserDefaults().fontForKey("Font: 2") {
        didSet {
            tokens = highlightingTokens()
        }
    }
    
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.commonSetUp()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        self.commonSetUp()
    }
    
    
    func commonSetUp() {
        self.font = defaultFont
        self.textColor = UIColor.whiteColor()
        self.indicatorStyle = .White
        
        self.tokens = highlightingTokens()
    }
    
    func highlightingTokens() -> [CYRToken!] {
        
        let regexAttributes = [
            NSForegroundColorAttributeName : SyntaxHighlighterDefaultColors.darkGoldColor
        ]
        
        let commentAttributes = [
                NSForegroundColorAttributeName : SyntaxHighlighterDefaultColors.commentGreen,
        ]
    
        
        
        let tokens = [
            
//            CYRToken(name: "special_numbers",
//                expression: "[ʝ]",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.orangeColor()
//                ]
//            ),
//            
//            CYRToken(name: "mod",
//                expression: "\\bmod\\b",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 245, g: 0, b: 110)
//                ]
//            ),
//
//            
//            CYRToken(name: "hex_1",
//                expression: "\\$[\\d a-f]+",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 186, g: 114, b: 181)
//                ]
//            ),
//            
//            CYRToken(name: "octal_1",
//                expression: "&[0-7]+",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 186, g: 114, b: 181)
//                ]
//            ),
//            
//            CYRToken(name: "binary_1",
//                expression: "%[01]+",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 186, g: 114, b: 181)
//                ]
//            ),
//            
//            
//            CYRToken(name: "hex_2",
//                expression: "0x[0-9 a-f]+",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 186, g: 114, b: 181)
//                ]
//            ),
//            
//            
//            CYRToken(name: "octal_2",
//                expression: "0o[0-7]+",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 186, g: 114, b: 181)
//                ]
//            ),
//            
//            
//            CYRToken(name: "binary_2",
//                expression: "0b[01]+",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 186, g: 114, b: 181)
//                ]
//            ),
//            
//            
//            CYRToken(name: "float",
//                expression: "\\d+\\.?\\d+e[\\+\\-]?\\d+|\\d+\\.\\d+|∞",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 186, g: 114, b: 181)
//                ]
//            ),
//            
//            
//            CYRToken(name: "integer",
//                expression: "\\d+",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 186, g: 114, b: 181)
//                ]
//            ),
//            
//
//            CYRToken(name: "operator",
//                expression: "[/\\*,\\;:=<>\\+\\-\\^!·≤≥]",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 255, g: 255, b: 255)
//                ]
//            ),
//            
//            
//            CYRToken(name: "round_brackets",
//                expression: "[\\(\\)]",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 255, g: 255, b: 255)
//                ]
//            ),
//            
//            CYRToken(name: "square_brackets",
//                expression: "[\\[\\]]",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 255, g: 255, b: 255)
//                ]
//            ),
//            
//            CYRToken(name: "absolute_brackets",
//                expression: "[|]",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 255, g: 255, b: 171)
//                ]
//            ),
            
            CYRToken(name: "reserved_words",
                expression: "\\b(as|async|await|break|case|catch|class|const|continue|debugger|default|delete|do|else|enum|export|extends|finally|for|from|function|get|if|implements|import|in|instanceof|interface|let|new|null|of|package|private|protected|public|return|set|static|super|switch|this|throw|try|typeof|var|void|while|with|yield)\\b",
                attributes: HighlighterExtention.attributesForKey(.Tag)
            ),

            
//            CYRToken(name: "chart_parameters",
//                expression: "(chartheight|charttitle|chartwidth|color|seriesname|showlegend|showxmajorgrid|showxminorgrid|showymajorgrid|showyminorgrid|transparency|thickness|xautoscale|xaxisrange|xlabel|xlogscale|xrange|yautoscale|yaxisrange|ylabel|ylogscale|yrange)",
//                attributes:
//                [
//                    NSForegroundColorAttributeName : UIColor.rgb(r: 186, g: 114, b: 181)
//                ]
//            ),
            
            CYRToken(name: "string",
                expression: "\".*?(\"|$)|'.*?('|$)",
                attributes: HighlighterExtention.attributesForKey(.String)
            ),
            
            
            
            CYRToken(name: "regular_expression",
                expression: "/.*?(/|$)",
                attributes: regexAttributes
            ),
            
            CYRToken(name: "regular_expression_with_I",
                expression: "/.*?(/i|$)",
                attributes: regexAttributes
            ),
            
            CYRToken(name: "regular_expression_with_G",
                expression: "/.*?(/g|$)",
                attributes: regexAttributes
            ),
            
            
            
            CYRToken(name: "documentation_comment",
                expression: "\\/\\*[\\w\\W]*?\\*\\/",
                attributes: commentAttributes
            ),
            
            
            
            CYRToken(name: "comment",
                expression: "(^|[^\\\\:])\\/\\/.*",
                attributes: commentAttributes
            )
            
        ]
        
        return tokens
    }
}
